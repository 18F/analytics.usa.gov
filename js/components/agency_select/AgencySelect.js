import React from "react";
import PropTypes from "prop-types";

/**
 * Creates a select tag with options populated from the provided props. When an
 * option is chosen, the browser is redirected to the page for that agency.
 *
 * @param {object} props the properties for the component
 * @param {string} props.mainAgencyName the option name to display for the
 * default option in the select.
 * @param {string} props.agencies a JSON string of an array of objects with slug
 * and name keys. 'slug' is the page path for the agency, and 'name' is the
 * display name for the option.
 * @param {string} props.pathSuffix a URL path to append to the end of the
 * agency slug.
 * @returns {import('react').ReactElement} The rendered element
 */
function AgencySelect({ mainAgencyName, agencies, pathSuffix = "" }) {
  function onChange(event) {
    window.location.assign(
      new URL(window.location.origin + event.target.selectedOptions[0].value),
    );
  }

  const options = nestAgencies(parseAgencies(agencies, mainAgencyName));
  const currentPath = window.location.pathname;
  const defaultValue =
    currentPath.slice(-1) == "/" ? currentPath.slice(0, -1) : currentPath;

  function parseAgencies(agencies, mainAgencyName) {
    const parsedAgencies = JSON.parse(agencies);
    return [{ slug: "", name: mainAgencyName }, ...parsedAgencies];
  }

  function nestAgencies(agencies) {
    const nestChildren = (parent, list) => {
      parent.children = list.filter((x) => x.parent == parent.slug);
    };

    const topLevel = agencies.filter((x) => x.parent === undefined);
    topLevel.forEach((top) => nestChildren(top, agencies));
    return topLevel;
  }

  return (
    <>
      <label
        className="usa-label text-center margin-y-1"
        htmlFor="analytics-agency-select"
      >
        Select an agency to display results
      </label>
      <select
        aria-label="Select an agency to display results"
        className="usa-select"
        name="agency-selector"
        id="analytics-agency-select"
        title="Agency Selection Dropdown"
        onChange={onChange}
        defaultValue={defaultValue}
      >
        {options.map((option) => {
          let value = option.slug
            ? `/${option.slug}${pathSuffix}`
            : `${pathSuffix}`;

          return (
            <React.Fragment key={option.slug}>
              <option key={option.slug} value={value}>
                {option.name}
              </option>
              {createChildrenOptGroup(option)}
            </React.Fragment>
          );
        })}
      </select>
    </>
  );

  function createChildrenOptGroup(option) {
    if (option.children.length === 0) {
      return;
    }

    return (
      <optgroup
        key={`${option.slug}-children`}
        label={`${option.redirect_from[0].toUpperCase()} Sub-agencies`}
      >
        {option.children.map((child) => {
          return (
            <option key={child.slug} value={`/${child.slug}${pathSuffix}`}>
              {child.name}
            </option>
          );
        })}
      </optgroup>
    );
  }
}

AgencySelect.propTypes = {
  mainAgencyName: PropTypes.string.isRequired,
  agencies: PropTypes.string,
  pathSuffix: PropTypes.string,
};

export default AgencySelect;
