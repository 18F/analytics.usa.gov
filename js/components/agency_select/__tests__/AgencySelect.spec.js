import React from "react";
import renderer from "react-test-renderer";
import AgencySelect from "../AgencySelect";

describe("AgencySelect", () => {
  describe("when all params are provided", () => {
    describe("and subagencies exist", () => {
      it("renders", () => {
        const component = renderer.create(
          <AgencySelect
            pathSuffix="/data"
            mainAgencyName="foobar"
            agencies='[{"name":"Agency for International Development","slug":"agency-international-development","_name":"Agency for International Development"},{"name": "Department of the Interior","slug": "interior","redirect_from":["doi"]},{"name": "National Park Service","slug": "national-park-service","parent": "interior"}]'
          />,
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });

    describe("and subagencies do not exist", () => {
      it("renders", () => {
        const component = renderer.create(
          <AgencySelect
            pathSuffix="/data"
            mainAgencyName="foobar"
            agencies='[{"name":"Agency for International Development","slug":"agency-international-development","_name":"Agency for International Development"}]'
          />,
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });

  describe("when optional params are not provided", () => {
    it("renders", () => {
      const component = renderer.create(
        <AgencySelect
          pathSuffix=""
          mainAgencyName="foobar"
          agencies='[{"name":"Agency for International Development","slug":"agency-international-development","_name":"Agency for International Development"}]'
        />,
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
