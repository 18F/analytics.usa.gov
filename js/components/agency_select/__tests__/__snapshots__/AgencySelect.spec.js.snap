// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AgencySelect when all params are provided and subagencies do not exist renders 1`] = `
[
  <label
    className="usa-label margin-y-1"
    htmlFor="analytics-agency-select"
  >
    Select an agency to display results
  </label>,
  <select
    aria-label="Select an agency to display results"
    className="usa-select"
    defaultValue=""
    id="analytics-agency-select"
    name="agency-selector"
    onChange={[Function]}
    title="Agency Selection Dropdown"
  >
    <option
      value="/data"
    >
      foobar
    </option>
    <option
      value="/agency-international-development/data"
    >
      Agency for International Development
    </option>
  </select>,
]
`;

exports[`AgencySelect when all params are provided and subagencies exist renders 1`] = `
[
  <label
    className="usa-label margin-y-1"
    htmlFor="analytics-agency-select"
  >
    Select an agency to display results
  </label>,
  <select
    aria-label="Select an agency to display results"
    className="usa-select"
    defaultValue=""
    id="analytics-agency-select"
    name="agency-selector"
    onChange={[Function]}
    title="Agency Selection Dropdown"
  >
    <option
      value="/data"
    >
      foobar
    </option>
    <option
      value="/agency-international-development/data"
    >
      Agency for International Development
    </option>
    <option
      value="/interior/data"
    >
      Department of the Interior
    </option>
    <optgroup
      label="DOI Sub-agencies"
    >
      <option
        value="/national-park-service/data"
      >
        National Park Service
      </option>
    </optgroup>
  </select>,
]
`;

exports[`AgencySelect when optional params are not provided renders 1`] = `
[
  <label
    className="usa-label margin-y-1"
    htmlFor="analytics-agency-select"
  >
    Select an agency to display results
  </label>,
  <select
    aria-label="Select an agency to display results"
    className="usa-select"
    defaultValue=""
    id="analytics-agency-select"
    name="agency-selector"
    onChange={[Function]}
    title="Agency Selection Dropdown"
  >
    <option
      value=""
    >
      foobar
    </option>
    <option
      value="/agency-international-development"
    >
      Agency for International Development
    </option>
  </select>,
]
`;
